<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [1]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data


--2020-07-23 04:07:28--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25936 (25K) [application/x-httpd-php]
Saving to: ‘imports-85.data’

imports-85.data     100%[===================>]  25.33K  --.-KB/s    in 0.07s   

2020-07-23 04:07:29 (380 KB/s) - ‘imports-85.data’ saved [25936/25936]



In [2]:
!head imports-85.data

3,?,alfa-romero,gas,std,two,convertible,rwd,front,88.60,168.80,64.10,48.80,2548,dohc,four,130,mpfi,3.47,2.68,9.00,111,5000,21,27,13495
3,?,alfa-romero,gas,std,two,convertible,rwd,front,88.60,168.80,64.10,48.80,2548,dohc,four,130,mpfi,3.47,2.68,9.00,111,5000,21,27,16500
1,?,alfa-romero,gas,std,two,hatchback,rwd,front,94.50,171.20,65.50,52.40,2823,ohcv,six,152,mpfi,2.68,3.47,9.00,154,5000,19,26,16500
2,164,audi,gas,std,four,sedan,fwd,front,99.80,176.60,66.20,54.30,2337,ohc,four,109,mpfi,3.19,3.40,10.00,102,5500,24,30,13950
2,164,audi,gas,std,four,sedan,4wd,front,99.40,176.60,66.40,54.30,2824,ohc,five,136,mpfi,3.19,3.40,8.00,115,5500,18,22,17450
2,?,audi,gas,std,two,sedan,fwd,front,99.80,177.30,66.30,53.10,2507,ohc,five,136,mpfi,3.19,3.40,8.50,110,5500,19,25,15250
1,158,audi,gas,std,four,sedan,fwd,front,105.80,192.70,71.40,55.70,2844,ohc,five,136,mpfi,3.19,3.40,8.50,110,5500,19,25,17710
1,?,audi,gas,std,four,wagon,fwd,front,105.80,192.70,71.40,55.70,2954,ohc,five,136,mpfi,3.19,3.40,8.50,1

In [3]:
import pandas as pd
import numpy as np
import scipy.stats as stats

In [4]:
columns = ['symboling',
           'normalized-losses',
           'make',
           'fuel-type',
           'aspiration',
           'num-of-doors',
           'body-style',
           'drive-wheels',
           'engine-location',
           'wheel-base',
           'length',
           'width',
           'height',
           'curb-weight',
           'engine-type',
           'num-of-cylinders',
           'engine-size',
           'fuel-system',
           'bore',
           'stroke'
           'compression-ratio',
           'horsepower',
           'peak-rpm',
           'city-mpg',
           'highway-mpg',
           'price']

In [5]:
cars = pd.read_csv('imports-85.data', header=None, names=columns, na_values='?')

In [6]:
cars.head()

symboling normalized-losses make  ... city-mpg highway-mpg    price
3        NaN       alfa-romero  gas  ...       21          27  13495.0
3        NaN       alfa-romero  gas  ...       21          27  16500.0
1        NaN       alfa-romero  gas  ...       19          26  16500.0
2      164.0              audi  gas  ...       24          30  13950.0
2      164.0              audi  gas  ...       18          22  17450.0

[5 rows x 25 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [7]:
cars_samp = cars.sample(n=10, random_state=30)

In [8]:
hwympg_pop = cars['highway-mpg'].mean()
hwympg_pop

30.75121951219512

In [9]:
stats.ttest_1samp(cars_samp['highway-mpg'], hwympg_pop, nan_policy='omit')

Ttest_1sampResult(statistic=0.8912422391619923, pvalue=0.39600040771093326)

Based on a t-statistic of 0.89 and a pvalue of 0.39, the sample mean is not the same as but similar to the population mean.

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [10]:
stats.ttest_1samp(cars_samp['highway-mpg'], 35, nan_policy='omit')

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

The salesman is the epitome of a sterotypical car salesman. Based on a t-statistic of -0.96 and a pvalue of 0.36, the sample mean is somewhat below the salesman's claim of 35 miles per gallon highway fuel efficiency. 

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [11]:
cars_samp2 = cars.sample(100, random_state=30)

In [12]:
stats.ttest_1samp(cars_samp2['highway-mpg'], 35, nan_policy='omit')

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

According to a t-statistic of -6.52 and a pvalue of 3.003^-9, the actual fuel efficiency of the vehicles tested is below the grossly overstated 35 mpg.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

With the smaller sample size, the possibility exists for the sample set to include cars with the top ten best or worst fuel economy. This would drastically skew the results one direction or the other leading to misrepresentation of the population.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [13]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-07-23 04:07:31--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.07s   

2020-07-23 04:07:31 (271 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [14]:
!head house-votes-84.data

republican,n,y,n,y,y,y,n,n,n,y,?,y,y,y,n,y
republican,n,y,n,y,y,y,n,n,n,n,n,y,y,y,n,?
democrat,?,y,y,?,y,y,n,n,n,n,y,n,y,y,n,n
democrat,n,y,y,n,?,y,n,n,n,n,y,n,y,n,n,y
democrat,y,y,y,n,y,y,n,n,n,n,y,?,y,y,y,y
democrat,n,y,y,n,y,y,n,n,n,n,n,n,y,y,y,y
democrat,n,y,n,y,y,y,n,n,n,n,n,n,?,y,y,y
republican,n,y,n,y,y,y,n,n,n,n,n,n,y,y,?,y
republican,n,y,n,y,y,y,n,n,n,n,n,y,y,y,n,y
democrat,y,y,y,n,n,n,y,y,y,n,n,n,n,n,?,?


In [15]:
vote_columns = ['Class-Name',
                'handicapped-infants',
                'water-projec-cost-sharing',
                'adoption-of-the-budget-resolution',
                'physician-fee-freeze',
                'el-salvador-aid',
                'religious-froups-in-schools',
                'anti-satellite-test-ban',
                'aid-to-nicaraguan-contras',
                'mx-missile',
                'immigration',
                'synfuels-corporation-cutback',
                'education-spending',
                'superfund-right-to-sue',
                'crime',
                'duty-free-exports',
                'export-administration-act-south-africa']

In [16]:
votes = pd.read_csv('house-votes-84.data', header=None, names=vote_columns, na_values='?')

In [17]:
votes.head()

Class-Name  ... export-administration-act-south-africa
0  republican  ...                                      y
1  republican  ...                                    NaN
2    democrat  ...                                      n
3    democrat  ...                                      y
4    democrat  ...                                      y

[5 rows x 17 columns]

In [18]:
votes.replace(to_replace=['y','n'], value=[1,0], inplace=True)

In [19]:
votes.head()

Class-Name  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [20]:
rep_votes = votes[votes['Class-Name']=='republican']
rep_votes.head()

Class-Name  ...  export-administration-act-south-africa
0   republican  ...                                     1.0
1   republican  ...                                     NaN
7   republican  ...                                     1.0
8   republican  ...                                     1.0
10  republican  ...                                     0.0

[5 rows x 17 columns]

In [21]:
dem_votes = votes[votes['Class-Name']=='democrat']
dem_votes.head()

Class-Name  ...  export-administration-act-south-africa
2   democrat  ...                                     0.0
3   democrat  ...                                     1.0
4   democrat  ...                                     1.0
5   democrat  ...                                     1.0
6   democrat  ...                                     1.0

[5 rows x 17 columns]

In [22]:
stats.ttest_ind(rep_votes['handicapped-infants'], dem_votes['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)

Based on a t-statistic of -9.21 and a p-value of 1.61^-18, republicans voted against the handicapped infants issue while democrats voted for it.

In [23]:
stats.ttest_ind( dem_votes['anti-satellite-test-ban'], rep_votes['anti-satellite-test-ban'], nan_policy='omit')

Ttest_indResult(statistic=12.526187929077842, pvalue=8.521033017443867e-31)

Base on a t-statistic of 12.53 and p-value of 8.52^-31, democrats voted for the anti-satellite-test-ban issue while republicans voted against it.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [24]:
rep_votes.head()

Class-Name  ...  export-administration-act-south-africa
0   republican  ...                                     1.0
1   republican  ...                                     NaN
7   republican  ...                                     1.0
8   republican  ...                                     1.0
10  republican  ...                                     0.0

[5 rows x 17 columns]

In [25]:
votes.columns

Index(['Class-Name', 'handicapped-infants', 'water-projec-cost-sharing',
       'adoption-of-the-budget-resolution', 'physician-fee-freeze',
       'el-salvador-aid', 'religious-froups-in-schools',
       'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile',
       'immigration', 'synfuels-corporation-cutback', 'education-spending',
       'superfund-right-to-sue', 'crime', 'duty-free-exports',
       'export-administration-act-south-africa'],
      dtype='object')

In [26]:
for i in votes.columns:
    if votes.columns.get_loc(i) > 0:
        print(i, ': ', stats.ttest_ind(rep_votes[i], dem_votes[i], nan_policy='omit'))
        print('\n')


handicapped-infants :  Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)


water-projec-cost-sharing :  Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)


adoption-of-the-budget-resolution :  Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)


physician-fee-freeze :  Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)


el-salvador-aid :  Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68)


religious-froups-in-schools :  Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)


anti-satellite-test-ban :  Ttest_indResult(statistic=-12.526187929077842, pvalue=8.521033017443867e-31)


aid-to-nicaraguan-contras :  Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54)


mx-missile :  Ttest_indResult(statistic=-16.437503268542994, pvalue=5.03079265310811e-47)


immigration :  Ttest_indResult(statistic=1.7359117329695164, pvalu



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [27]:
import statistics as st

In [28]:
#select data
rep_votes_manual = rep_votes['crime'].dropna()

In [29]:
#get sample mean
rep_manual = rep_votes_manual
rep_manual_mean = rep_manual.mean()

In [31]:
#get total population
pop_manual = votes['crime'].sample(n=round(len(votes['crime'].index)*.2), random_state=20)
pop_manual_mean = pop_manual.mean()

In [32]:
#sample size
samp_size = len(rep_votes_manual-1)
samp_size

161

In [33]:
#standard deviation of sample population...statistics module needed
std_dev_samp_rep = st.stdev(rep_votes_manual)

In [34]:
#calculate t-statistic
t1 = (rep_manual_mean - pop_manual_mean) / (std_dev_samp_rep / samp_size**0.5)

In [35]:
p1 = -1.69913
print(t1, '\n', p1)

39.69872962775616 
 -1.69913


In [36]:
stats.ttest_1samp(rep_votes_manual, pop_manual.mean(), nan_policy='omit')

Ttest_1sampResult(statistic=39.69872962775616, pvalue=9.677541559392915e-85)

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE